### PR TITLE
Amy Cash - Pipes - Api-Muncher

### PR DESCRIPTION
**Pull request by cashmonger**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did you go about exploring the Edamam API, how did you try querying the API? |I cut and pasted the sample search and then changed parts to see what I would get. I tried searching for an individual recipe (r search) and searching by topic (q) search.  |
| Describe your API Wrapper.  How did you decide on the methods you created? | They were based on what we did with the Slack Api. The trickiest part was writing the method to show an individual recipe. |
| Describe an edge case or failure case test you wrote for your API Wrapper. | I tested for a blank search term with both the get_recipes and show_recipe methods.  |
| Explain how VCR aids in testing an API. |  It keeps you from having to query the API over and over again, which is faster, and, potentially cheaper, if you're paying for the calls. |
| What is the Heroku URL of your deployed application? | http://especiallyrecipes.herokuapp.com/ |

### CODE REVIEW

File: lib/api_muncher_wrapper.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    return recipe_list
+  end #method
+
+  def self.show_recipe(id)
+    puts "Searching for specific recipe"
+    puts "#{id}"
+    search_uri = "http://www.edamam.com/ontologies/edamam.owl" + "\#" + "#{id}"
+    url = BASE_URL + "search?r=#{URI.encode(search_uri)}"
+    data = HTTParty.get(url)
+    result = create_recipe(data[0])
```
Comment (droberts-sea) [inline lib/api_muncher_wrapper.rb:28]: You don't have any error handling here! You should be checking the messages sent back by the API, as well as ensuring it actually sent the data you want.

File: test/lib/api_muncher_wrapper_test.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        result = ApiMuncherWrapper.show_recipe("recipe_421df807e21c65c842ec62870604aed3")
+        result.must_be_kind_of Recipe
+      end
+    end
+
+    it "Raises an error when there is an invalid or blank URI" do
+      VCR.use_cassette("individual_recipe") do
+        proc {
+          ApiMuncherWrapper.show_recipe()
+        }.must_raise ArgumentError
```
Comment (droberts-sea) [inline test/lib/api_muncher_wrapper_test.rb:43]: This isn't testing the API wrapper itself - because `id` is a required argument for the `show_recipe` method, Ruby will automatically raise this error if you don't pass it. There's no line of code you could remove to make this assertion fail.

A more interesting test case might be something like

```ruby
recipe = ApiMuncherWrapper.show_recipe("bogus_recipe_uri")
recipe.must_be_nil
```

File: test/lib/api_muncher_wrapper_test.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+        proc {
+          ApiMuncherWrapper.get_recipes("123456")
+        }.must_raise ArgumentError
```
Comment (droberts-sea) [inline test/lib/api_muncher_wrapper_test.rb:47]: You should probably be calling `show_recipe` here, since that's what this test claims to be doing.

File: test/lib/api_muncher_wrapper_test.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    end #get list
+  end #get recipies
+
+  it "Raises an error when there is no search term" do
+    VCR.use_cassette("recipes") do
+      food = "peach"
+      from = 0
+      proc {
+        ApiMuncherWrapper.get_recipes(food)
+      }.must_raise ArgumentError
```
Comment (droberts-sea) [inline test/lib/api_muncher_wrapper_test.rb:23]: Again, this is testing that Ruby throws an error when you don't pass enough arguments, not any of the code you wrote. A more interesting test might be something like:

```ruby
results = ApiMuncherWrapper.get_recipes("this_search_term_shouldnt_have_any_results", 0)
results.must_be_kind_of Array
results.must_be :empty?
# Or check that it throws an error, if that's the behavior you want
```

Other test cases I'd be interested in:

- What if the `from` parameter is past the end of the search results? What about if there are less than 10 results remaining?
- What if the `search` parameter is the empty string?

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
| **Core Requirements** |   |
| Git hygiene | yes |
| Comprehension questions	| yes |
| **General** |  |
| Rails fundamentals (RESTful routing, use of named paths) | yes |
| Semantic HTML | some - There are lots of `<div>`s where you could use `<article>` or `<section>`. |
| Errors are reported to the user | some - For search they are reported nicely, but for show if I mangle the URI I get "We're sorry, but something went wrong". It looks like there's a disconnect in your error handling - see inline comment. |
| API Wrapper to handle the API requests | yes |
| Controller testing | no |
| Lib testing | Success cases look good, but there's some work to do around failure cases. |
| Search Functionality | yes |
| List Functionality | yes |
| Show individual item functionality (link to original recipe opens in new tab) | yes |
| **Styling** |  |
| Foundation Styling for responsive layout | yes |
| List View shows 10 items at a time/pagination | yes |
| The app is styled to create an attractive user interface | yes |
| **API Features** | |
| The App attributes Edaman | yes |
| The VCR casettes do not contain the API key | N/A - From conversation in class, you were able to complete the assignment without keys. Make sure you understand why this is an important requirement. |
| **External Resources** |  |
| Link to deployed app on Heroku | yes |
|  **Overall** | This is a good start. You've got an attractive and functional website that consumes the Edamam API, and it seems like most of the core learning goals were met.<br><br>I am a little concerned about your lack of test coverage, and the difficulty you had connecting some of the error handling. These two skills are closely linked - they both require you to be able to take a step back and think about your method as a unit: what does success or failure look like, what combinations of inputs might cause it to fail, and what should I do when it does fail. Writing code that's stable and resilient is not easy, but it's what separates a good application from a great one. Make sure you spend some time thinking about error handling and testing with VideoStore API, and let us know if you need more support on this. |
